src/_macosx.m

```diff
@@ -41,6 +41,66 @@
 static bool keyChangeCapsLock = false;
 /* Keep track of the current mouse up/down state for open/closed cursor hand */
 static bool leftMouseGrabbing = false;
+/* Keep track of whether stdin has been received */
+bool stdin_received = false;
+bool stdin_sigint = false;
+
+// Signal handler for SIGINT
+void handleSigint(int signal) {
+    stdin_sigint = true;
+}
+
+int wait_for_stdin() {
+    @autoreleasepool {
+        stdin_received = false;
+        stdin_sigint = false;
+
+        // Set up a SIGINT handler to interrupt the event loop if ctrl+c comes in too
+        struct sigaction action = {0};
+        action.sa_handler = handleSigint;
+        sigaction(SIGINT, &action, NULL);
```

Comment:
Do we need to check for an existing signal handler (set by Python for example)?

Reply:
I added that capability in, and now the code requires two ctrl+c to exit. The first one only exits the event loop, but then we are still waiting for the `input` block I think so it needs either an `enter` or `ctrl+c` to kill that one again. Keeping the `raise SIGINT` in there didn't matter one way or the other.

I'm not sure what is preferable. Setting it to the default sigint handler kills the process immediately. But I suppose this is OK as long as you know you need to kills sent to exit? I am not sure about how this is typically handled so any advice/suggestions are welcome.

Reply:
I would guess something similar to the SIGINT handling in Qt?

Reply:
Is there a reference implementation for that? I'm still not following what the expected/desired path is compared to what we currently have.

Just to clarify the current situation with that example code and then going to the terminal that is waiting for `[enter]` input.

`macosx`: `ctrl+c` -> leaves the event loop, then waits for the input still. Another `ctrl+c` exits the program with a `KeyboardInterrupt`.
`qtagg`: `ctrl+c` -> does nothing while waiting for the input, so hitting it multiple times does not kill the program. Clicking on the figure window again (bouncing the event loop I guess?) kills the program ungracefully with an abort.
`tkagg`: `ctrl+c` -> kills the program with a fatal Python error
```
Fatal Python error: PyEval_RestoreThread: the function must be called with the GIL held, but the GIL is released (the current Python thread state is NULL)
Python runtime state: initialized
```

Out of all of these, I'd say that the `macosx` state in this PR is the most ideal...

Reply:
@anntzer pinging you here in case you have thoughts on this PR and the current status

Reply:
Noted, but I cannot guarantee any amount of time to look at this in depth at least until the end of the week (for sure) and possibly for even longer.